s3ninja appears to require signed chunks for PUT

Hi. Thanks for s3ninja, it's been really helpful for us!

We're encountering a problem with pre-signed S3 URLs. We're attempting to do a plain old HTTP PUT on a pre-signed URL (not using the AWS SDK) with the body data `foo`. When we do this, we see an error in the s3ninja log and the PUT fails:

```
Caused by: java.lang.NumberFormatException: For input string: "foo"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at ninja.SignedChunkHandler.handle(SignedChunkHandler.java:55)
	at sirius.web.http.WebServerHandler.channelReadHttpContent(WebServerHandler.java:310)
	at sirius.web.http.WebServerHandler.channelReadLastHttpContent(WebServerHandler.java:320)
	at sirius.web.http.WebServerHandler.channelRead(WebServerHandler.java:275)
```

As far as I can tell from the code, it looks like s3ninja assumes that the request will always be chunked and that the chunks will always be signed by the client, however in this case the client is doing a plain, non-chunked PUT request to a pre-signed URL.

This used to work when creating the pre-signed URL with AWS SDK 1.10.x (so AWS v2 signing) and s3ninja 2.7. Now we need to upgrade our system to use AWS SDK 1.11.x for other reasons, so our generated pre-signed URLs are now using AWS v4 signing, and so we need to upgrade s3ninja (because 2.7 will not support v4 signing). Unfortunately s3ninja versions _after_ 2.7 no longer support a plain PUT (chunks without signatures) to a pre-signed URL.

I've tried every version 4, 5, 5.1, 5.2, 5.2.1, 5.2.2, 6 and the current HEAD (bd64601de4032b770e1f94e2dd38f064e4b46926).

--- a/src/main/java/ninja/S3Dispatcher.java
+++ b/src/main/java/ninja/S3Dispatcher.java
@@ -142,7 +142,7 @@ public class S3Dispatcher implements WebDispatcher {
     }
 
     private InputStreamHandler createInputStreamHandler(WebContext ctx) {
-        if (aws4HashCalculator.supports(ctx) && ctx.getRequest().method() == PUT) {
+        if (aws4HashCalculator.supports(ctx) && ctx.getRequest().method() == PUT && ctx.getHeader("x-amz-decoded-content-length") != null) {
             return new SignedChunkHandler();
         } else {
             return new InputStreamHandler();

--- a/src/test/java/BaseAWSSpec.groovy
+++ b/src/test/java/BaseAWSSpec.groovy
@@ -5,6 +5,7 @@
  * Copyright by scireum GmbH
  * http://www.scireum.de - info@scireum.de
  */
+import com.amazonaws.HttpMethod
 import com.amazonaws.services.s3.AmazonS3Client
 import com.amazonaws.services.s3.model.AmazonS3Exception
 import com.amazonaws.services.s3.model.GeneratePresignedUrlRequest
@@ -159,4 +160,35 @@ abstract class BaseAWSSpec extends BaseSpecification {
         AmazonS3Exception e = thrown()
         e.statusCode == 404
     }
+
+    def "PUT on presigned URL without signed chunks works as expected"() {
+        given:
+        def client = getClient()
+        when:
+        if (!client.doesBucketExist("test")) {
+            client.createBucket("test")
+        }
+        and:
+        def content = "NotSigned"
+        and:
+        GeneratePresignedUrlRequest putRequest = new GeneratePresignedUrlRequest("test", "test", HttpMethod.PUT)
+        HttpURLConnection hc = new URL(getClient().generatePresignedUrl(putRequest).toString()).openConnection()
+        hc.setDoOutput(true)
+        hc.setRequestMethod("PUT")
+        OutputStreamWriter out = new OutputStreamWriter(hc.getOutputStream())
+        try {
+            out.write(content)
+        } finally {
+            out.close()
+        }
+        hc.getResponseCode()
+        and:
+        GeneratePresignedUrlRequest request = new GeneratePresignedUrlRequest("test", "test")
+        URLConnection c = new URL(getClient().generatePresignedUrl(request).toString()).openConnection()
+        and:
+        String downloadedData = new String(ByteStreams.toByteArray(c.getInputStream()), Charsets.UTF_8)
+        then:
+        downloadedData == content
+    }
+
 }

